src/plugins/audioplayers/player.c

```diff
@@ -0,0 +1,413 @@
+            GstState old_state, new_state;
+
+            gst_message_parse_state_changed(message, &old_state, &new_state, NULL);
+            AudioPlayer_OnMediaStateChange(data, message->src, &old_state, &new_state);
+            break;
+        }
+        case GST_MESSAGE_EOS:
+            gst_element_set_state(data->playbin, GST_STATE_READY);
+            AudioPlayer_OnPlaybackEnded(data);
+            break;
+        case GST_MESSAGE_DURATION_CHANGED:
+            AudioPlayer_OnDurationUpdate(data);
+            break;
+        case GST_MESSAGE_ASYNC_DONE:
+            if (!data->_isSeekCompleted) {
+                AudioPlayer_OnSeekCompleted(data);
+                data->_isSeekCompleted = true;
+            }
+            break;
+        default:
+            // For more GstMessage types see:
+            // https://gstreamer.freedesktop.org/documentation/gstreamer/gstmessage.html?gi-language=c#enumerations
+            break;
+    }
+
+    // Continue watching for messages
+    return TRUE;
+}
+
+gboolean AudioPlayer_OnRefresh(AudioPlayer* data)
+{
+    if (data->playbin->current_state == GST_STATE_PLAYING) {
+        AudioPlayer_OnPositionUpdate(data);
+    }
+    return TRUE;
+}
+
+void AudioPlayer_SetPlayback(AudioPlayer* self, int64_t seekTo, double rate)
+{
+    if (!self->_isInitialized) {
+        return;
+    }
+    // See:
+    // https://gstreamer.freedesktop.org/documentation/tutorials/basic/playback-speed.html?gi-language=c
+    if (!self->_isSeekCompleted) {
+        return;
+    }
+    if (rate == 0) {
+        // Do not set rate if it's 0, rather pause.
+        AudioPlayer_Pause(self);
+        return;
+    }
+
+    if (self->_playbackRate != rate) {
+        self->_playbackRate = rate;
+    }
+    self->_isSeekCompleted = false;
+
+    GstSeekFlags seek_flags = GST_SEEK_FLAG_FLUSH | GST_SEEK_FLAG_ACCURATE;
+    GstEvent* seek_event;
+    if (rate > 0) {
+        seek_event =
+          gst_event_new_seek(rate, GST_FORMAT_TIME, seek_flags, GST_SEEK_TYPE_SET, seekTo * GST_MSECOND, GST_SEEK_TYPE_NONE, -1);
+    } else {
+        seek_event =
+          gst_event_new_seek(rate, GST_FORMAT_TIME, seek_flags, GST_SEEK_TYPE_SET, 0, GST_SEEK_TYPE_SET, seekTo * GST_MSECOND);
+    }
+    if (!gst_element_send_event(self->playbin, seek_event)) {
+        // FIXME
+        LOG_ERROR("Could not set playback to position %ld and rate %f.", seekTo, rate);
+        self->_isSeekCompleted = true;
+    }
+}
+void AudioPlayer_OnMediaError(AudioPlayer* self, GError* error, gchar* debug)
+{
+    (void)debug;
+    char* error_message = malloc(256 * sizeof(char));
+    snprintf(error_message, 256, "Error: %d; message=%s", error->code, error->message);
+    LOG_ERROR("%s", error_message);
+    if (self->_channel) {
+        platch_call_std(
+          self->_channel,
+          "audio.onError",
+          &STDMAP2(STDSTRING("playerId"), STDSTRING(self->playerId), STDSTRING("value"), STDSTRING(error_message)),
+          NULL,
+          NULL);
```

Comment:
```suggestion
        platch_call_std(
            self->_channel,
            "audio.onError",
            &STDMAP2(
                STDSTRING("playerId"), STDSTRING(self->playerId),
                STDSTRING("value"), STDSTRING(error_message)
            ),
            NULL,
            NULL
        );
```

Reply:
I prefer the way it looks in dart here, e.g. if it doesn't fit in a single line then split it up like
```C
var = function(
    arg1,
    arg2,
    arg3,
    arg4
);
```

And you can group values together on the same line if they fit together semantically somehow (like, the STDMAP2 key and values). Clang-format will break it up again so you might want to add a `// clang-format off` and `// clang-format on` as well